### ISSUE TITLE
Divergence not working for second order tensors

### ISSUE THREAD

**Issue by termi-official** [Ferrite-FEM/Tensors.jl#209]
MWE

```julia
julia> Tensors.divergence(x->Tensor{2,2}((x[1],0.0,0.0,x[2])), Vec((1.0,1.0)))
ERROR: MethodError: no method matching _extract_gradient(::Tensor{2, 2, ForwardDiff.Dual{ForwardDiff.Tag{var"#55#56", Vec{2, Float64}}, Float64, 2}, 4}, ::Vec{2, Float64})

Closest candidates are:
  _extract_gradient(::AbstractTensor{<:Any, <:Any, <:ForwardDiff.Dual}, ::Number)
   @ Tensors ~/.julia/packages/Tensors/9MMoU/src/automatic_differentiation.jl:58
  _extract_gradient(::Vec{2, <:ForwardDiff.Dual}, ::Vec{2})
   @ Tensors ~/.julia/packages/Tensors/9MMoU/src/automatic_differentiation.jl:92
  _extract_gradient(::Tensor{2, 2, <:ForwardDiff.Dual}, ::Tensor{2, 2})
   @ Tensors ~/.julia/packages/Tensors/9MMoU/src/automatic_differentiation.jl:125
  ...

Stacktrace:
 [1] gradient(f::var"#55#56", v::Vec{2, Float64})
   @ Tensors ~/.julia/packages/Tensors/9MMoU/src/automatic_differentiation.jl:456
 [2] divergence(f::var"#55#56", v::Vec{2, Float64})
   @ Tensors ~/.julia/packages/Tensors/9MMoU/src/automatic_differentiation.jl:524
 [3] top-level scope
   @ REPL[46]:1

```

**Comment by KnutAM** [comment]
I don't think this is correct, and currently, the code is special-cased for vector fields. 

We should probably adopt the method from Ferrite: https://github.com/Ferrite-FEM/Ferrite.jl/blob/fe44e7f6297158e9bbb62fd4bc9adaa1d79929f2/src/FEValues/common_values.jl#L107-L110

I.e., something like the following 
```julia
divergence(f::F, v::Vec) = divergence_from_gradient(gradient(f, v))
divergence_from_gradient(g::Vec) = sum(g)
divergence_from_gradient(g::SecondOrderTensor) = tr(g)
divergence_from_gradient(g::Tensor{3,dim,T}) where {dim,T} = g \boxdot one(Tensor{2,dim,T})
```

Btw. I also noticed that `((t::Tensor{3})[:,1,1])::Vector` which is not ideal, so this should also be fixed at some point!

**Comment by termi-official** [comment]
Good point, I guess the divergence of a tensor is in general not the trace of the gradient. :)
 
I think that your definition holds only for symmetric tensors $S$. Let us assume Cartesian coordinates, then 
$$div(S) = \nabla \cdot S = \partial_i S_{ik} e_k $$
where the first equal just uses the assumption of Cartesian spaces and the second equal uses the definitions of nabla and A+standard tensor algebra. Also in Cartesian coordinates we have
$$tr(grad(S)) = tr(\partial_k S_{ij} e_i \otimes e_j \otimes e_k) = \partial_i S_{ij} e_j $$
where the trace is used as above defined. Am I missing something else?

**Comment by KnutAM** [comment]
For the definition, yes these are the different definitions I was mentioning, either 
$$\mathrm{div}(\boldsymbol{S}) = \nabla \cdot \boldsymbol{S}$$ or $$\mathrm{div}(\boldsymbol{S}) = \boldsymbol{S} \cdot \nabla$$

Where the latter is used in, e.g. Bonet and Wood, but I've seen the one you refer to at other sources as well (such as the Wikipedia page for divergence of 2nd order tensors). 

For the second point, the difference between your code and the tensor expression is that the codes sum all entries, the "correct" code would be 
```julia
Vec{dim}(i->sum(t[k,i,l]*(k==l) for k in 1:dim, l in 1:dim))
```

(And for talking about definitions, we haven't started with the curl of 2nd order yet, I think I've seen 4 different definitions there :))

**Comment by KnutAM** [comment]
xref #210 for further discussions about divergence and curl definitions in general
